File: 15_3Sum.java
// Two Pointers
// Time complexity: O(n^2) for double loop, O(nlogn) for sorting, overall O(n^2).
// Space complexity: 
//      O(1) or O(n) extra space depending on the sorting algorithm.
//      O(m) space for the output list.

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        // 1. sort
        Arrays.sort(nums);

        // 2. outer loop i in [0, len-2)
        for (int i = 0; i < nums.length - 2; i++) {
            // 3. short-cut: avoid duplicates || max_sum < 0 || min_sum > 0
            if (i > 0 && nums[i] == nums[i-1]) continue;
            if (nums[i] + nums[nums.length-2] + nums[nums.length-1] < 0) continue;
            if (nums[i] + nums[i+1] + nums[i+2] > 0) break;

            // 4. 2-ptr for inner loop j, k
            int j = i+1, k = nums.length - 1;
            while (j < k) {
                int tmp = nums[i] + nums[j] + nums[k];
                if (tmp < 0) j += 1;
                else if (tmp > 0) k -= 1;
                else {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    // 5. avoid duplicates
                    j += 1;
                    k -= 1;
                    while (j < k && nums[j] == nums[j-1]) j += 1;
                    while (j < k && nums[k] == nums[k+1]) k -= 1;
                }
            }
        }

        return res;
    }
}

File: 15_3Sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        # 1. sort in ASC order
        nums.sort()
        # 2. for loop - 2Sum
        res = []
        for i in range(len(nums) - 2):
            # 3. avoid duplication & short-cut
            if i > 0 and nums[i] == nums[i-1]: continue
            if nums[i] + nums[-1] + nums[-2] < 0: continue
            if nums[i] + nums[i+1] + nums[i+2] > 0: break

            # 4. 2-ptr for sum-up-to 0
            l, r = i + 1, len(nums) - 1
            while l < r:
                tsum = nums[i] + nums[l] + nums[r]

                if tsum > 0: r -= 1
                elif tsum < 0: l += 1
                else:
                    res.append((nums[i], nums[l], nums[r]))
                    # 5. shift l/r ptr w/n duplicates
                    l, r = l + 1, r - 1
                    while l < r and nums[l] == nums[l-1]: l += 1
                    while l < r and nums[r] == nums[r+1]: r -= 1

        return res